odl/solvers/smooth/nonlinear_cg.py

```diff
@@ -0,0 +1,139 @@
+
+    Parameters
+    ----------
+    f : `Functional`
+        Functional with ``f.gradient``.
+    x : ``op.domain`` element
+        Vector to which the result is written. Its initial value is
+        used as starting point of the iteration, and its values are
+        updated in each iteration step.
+    line_search : float or `LineSearch`, optional
+        Strategy to choose the step length. If a float is given, uses it as a
+        fixed step length.
+    maxiter : int, optional
+        Maximum number of iterations to perform.
+    nreset : int, optional
+        Number of times the solver should be reset. Default: no reset.
+        fixed step length. Default: `BacktrackingLineSearch`
+    tol : float, optional
+        Tolerance that should be used for terminating the iteration.
+    beta_method : {'FR', 'PR', 'HS', 'DY'}
+        Method to calculate ``beta`` in the iterates.
+        * ``'FR'`` : Fletcher-Reeves
+        * ``'PR'`` : Polak-Ribiere
+        * ``'HS'`` : Hestenes-Stiefel
+        * ``'DY'`` : Dai-Yuan
+    callback : `callable`, optional
+        Object executing code per iteration, e.g. plotting each iterate
+
+    See Also
+    --------
+    bfgs_method : Quasi-newton solver for the same problem
+    conjugate_gradient : Optimized solver for linear and symmetric case
+    conjugate_gradient_normal : Equivalent solver but for linear case
+    """
+
+    if x not in f.domain:
+        raise TypeError('`x` {!r} is not in the domain of `f` {!r}'
+                        ''.format(x, f.domain))
+
+    if not callable(line_search):
+        line_search = ConstantLineSearch(line_search)
+
+    for _ in range(nreset + 1):
+        # First iteration is done without beta
+        dx = -f.gradient(x)
+        dir_derivative = -dx.inner(dx)
+        if abs(dir_derivative) < tol:
+            return
+        a = line_search(x, dx, dir_derivative)
+        x.lincomb(1, x, a, dx)  # x = x + a * dx
+
+        dx_old = dx
+        s = dx  # for 'HS' and 'DY' beta methods
+
+        for _ in range(maxiter // (nreset + 1)):
+            # Compute dx as -grad f
+            dx, dx_old = -f.gradient(x), dx
+
+            # Calculate "beta"
+            if beta_method == 'FR':
+                beta = dx.inner(dx) / dx_old.inner(dx_old)
+            elif beta_method == 'PR':
+                beta = dx.inner(dx - dx_old) / dx_old.inner(dx_old)
+            elif beta_method == 'HS':
+                beta = - dx.inner(dx - dx_old) / s.inner(dx - dx_old)
+            elif beta_method == 'DY':
+                beta = - dx.inner(dx) / s.inner(dx - dx_old)
+            else:
+                raise ValueError('unknown ``beta_method``')
+
+            # Reset beta if negative.
+            beta = max(0, beta)
+
+            # Update search direction
+            s.lincomb(1, dx, beta, s)  # s = dx + beta * s
+
+            # Find optimal step along s
+            dir_derivative = -dx.inner(s)
+            if abs(dir_derivative) < tol:
```

Comment:
Strictly speaking it should be `<=` here because "tolerance" in inclusive. It's also relevant if a user sets `tol=0` because in some (toy) cases the update could actually be 0 and the method  should terminate then.